lib/puppet/provider/pspackageprovider/powershellcore.rb

```diff
@@ -0,0 +1,89 @@
+require 'json'
+
+Puppet::Type.type(:pspackageprovider).provide :powershellcore do
+  confine operatingsystem: :windows
+  commands pwsh: 'pwsh'
+
+  mk_resource_methods
+
+  def self.invoke_ps_command(command)
+    # override_locale is necessary otherwise the Install-Module commands silently fails on Linux
+    result = Puppet::Util::Execution.execute(['pwsh', '-NoProfile', '-NonInteractive', '-NoLogo', '-Command',
+                                              "$ProgressPreference = 'SilentlyContinue'; $ErrorActionPreference = 'Stop'; #{command}"],
+                                             override_locale: false)
+    result.lines
+  end
+
+  def initialize(value = {})
+    super(value)
+    @property_flush = {}
+  end
+
+  def self.prefetch(resources)
+    instances.each do |prov|
+      if resource = resources[prov.name]
+        resource.provider = prov
+      end
+    end
+  end
+
+  def self.instances
+    result = invoke_ps_command instances_command
+    result.each.collect do |line|
+      p = JSON.parse(line.strip, symbolize_names: true)
+      p[:ensure] = :present
+      new(p)
+    end
+  end
+
+  def source_location=(value)
+    @property_flush[:source_location] = value
+  end
+
+  def exists?
+    @property_hash[:ensure] == :present
+  end
+
+  def create
+    self.class.invoke_ps_command install_command
+    @property_hash[:ensure] = :present
+  end
+
+  def flush
+    unless @property_flush.empty?
+      flush_command = "PackageManagement\\Install-PackageProvider -Name #{@resource[:name]}"
+      @property_flush.each do |key, value|
+        if @property_flush[:source_location]
+          flush_command << " -Source '#{value}'" 
+        elsif @property_flush[:version]
+          flush_command << " -RequiredVersion '#{value}'" 
+        else
+          flush_command << " -#{key} '#{value}'"
+        end
+      end
+      self.class.invoke_ps_command flush_command
```

Comment:
The flush_command will need a -Force on it otherwise you get a prompt when installing a new version

Reply:
D'oh! Good catch again!